### PR TITLE
Inv marker contour detection improved

### PR DESCRIPTION
**Pull request by szk1509**
### This pull request improves the accuracy of the white markers 

Resolves #2055.

- An accuracy improvement of the white marker corners was achieved.
- A slight improvement for the precision of the black markers corners was achieved.
- The achieved accuracy remains stable along different noises.

invertedMarkerDetection

This pull request modifies the contour selection for the white marker contours, by modifying the selection process of the corners and contour candidates, in __filterTooCloseCandidates_.

**candidatesSet**: The candidates and the contours are saved in a structur, for future improvements.

### CODE REVIEW

**Comment by alalek** [comment]
Please squash patch into one commit (remove unrelated commits from the patch and remove merge commits (use rebase instead of merge if necessary)).
After that rebase patch onto 3.4 branch: https://github.com/opencv/opencv/wiki/Branches

**Comment by szk1509** [comment]
Hi @alalek 

I have the Patch-Datei (i.e. the diff-datei with my changes) after that i have done a rebase to the 3.4 branch, then, if i use "git log" i can see the older commits  and if i use "git rebase -i" i can see only my current commit.
So i was wondering if i have to add and push the patch, or do the "git push --force origin <my-branch>" :thinking: which actually make sense, but i'm afraid of  ruin everything :D

thank you in advance for your help

**Comment by alalek** [comment]
I rebased PR changes onto 3.4 branch (please fetch commit from your fork, do not use merges, cherry-pick new commits if any).

**Comment by szk1509** [comment]
Thank you @alalek 
:thinking: fetch commits, use cherry pick and do not merge (use rebase instead of merge) are instructions for future PR?  or do i have to fetch, then commit and squash the commits if so?
Sorry!! i am a bit (too) confused :sob:

and again, thank you for your help :smiley:

File: modules/aruco/src/aruco.cpp [hunk 1]
```diff
@@ -523,9 +575,10 @@ static bool _identifyOneCandidate(const Ptr<Dictionary>& dictionary, InputArray
         if(invBError<borderErrors){
             borderErrors = invBError;
             invertedImg.copyTo(candidateBits);
+            typ=2;
```
Comment (alalek) [inline modules/aruco/src/aruco.cpp]: Please put some comment in code about meaning of these values.

File: modules/aruco/src/aruco.cpp [hunk 2]
```diff
@@ -212,17 +212,20 @@ static void _reorderCandidatesCorners(vector< vector< Point2f > > &candidates) {
 
 
 /**
-  * @brief Check candidates that are too close to each other and remove the smaller one
+  * @brief Check candidates that are too close to each other, save the potential candidates
+  *        (i.e. biggest/smallest contour) and remove the rest
   */
 static void _filterTooCloseCandidates(const vector< vector< Point2f > > &candidatesIn,
-                                      vector< vector< Point2f > > &candidatesOut,
+                                      vector< vector< vector< Point2f > > > &candidatesSetOut,
```
Comment (alalek) [inline modules/aruco/src/aruco.cpp]: > `vector< vector< vector<`

There are really two many nested vectors.
Two values are expected in outer dimension.

Perhaps we should split this parameter into two:
```
vector< vector< Point2f > > &candidatesOut,
vector< vector< Point2f > > &candidatesInvertedOut,
```

File: modules/aruco/src/aruco.cpp [hunk 3]
```diff
@@ -244,39 +247,86 @@ static void _filterTooCloseCandidates(const vector< vector< Point2f > > &candida
...
+
+            // check if current contour is smaller
+            if(currPerim < contoursIn[ smallerIdx ].size() && detectInvertedMarker)
+                smallerIdx = groupedCandidates[i][j];
+        }
+        // add contours und candidates
+        if(biggerIdx > -1){
+
+            biggerCandidates.push_back(candidatesIn[biggerIdx]);
+            biggerContours.push_back(contoursIn[biggerIdx]);
```
Comment (alalek) [inline modules/aruco/src/aruco.cpp]: if we observing lack of performance in this code, then we can try to utilize smart pointers:

```.diff
-const vector< vector< Point2f > > &candidatesIn
+typedef vector< Point2f > CandidatePoints;
+typedef cv::Ptr<CandidatePoints> CandidatePointsPtr;
+const vector<CandidatePointsPtr> &candidatesIn
```

**Comment by hamddan4** [comment]
I have a question regarding how it works, posted here: https://stackoverflow.com/questions/58302536/how-to-properly-detect-inverted-white-markers-with-opencv-aruco-module

**Comment by szk1509** [comment]
Hi @hamddan4,
The good news is that it works for me, so i think i know where the problem is :)
the bad news is that it comes soon on the next PR... 

if you like i can give you some more details, so that you can keep working ;)

**Comment by hamddan4** [comment]
Hi! Yes, i would like to know how to add this functionality to my code. Afaik, detectInvertedMarkers option of parameters is on the source code, but it doesn't seem to work.. what I am missing here?

**Comment by szk1509** [comment]
@ hamddan4 :thinking: well ...
The good news is that after trying 3.4 ad 4.1, it worked for me, so it works... my bad, I was in a hurry and couldn't check it properly.
The bad news is that I can't reproduce your error :thinking: so let's see...

import cv2
from cv2 import aruco
dic = aruco.getPredefinedDictionary(aruco.DICT_APRILTAG_16h5)
par = aruco.DetectorParameters_create()
par.detectInvertedMarker = True
image = cv2.imread("error/MNSGxZT.png")
dMarkers = aruco.detectMarkers( image, dic, parameters=par )
cv2.imshow("d", aruco.drawDetectedMarkers(image, dMarkers[0]))
cv2.waitKey(0)

i don't actually see what are you missing :thinking:

**Comment by hamddan4** [comment]
So if it works with you, it has to be some problem with the my installed libraries? Maybe the headers are good but is not implemented in the installed version of opencv. I've done this by `pip install opencv-contrib-python`. Don't know what's wrong with it though. Is there anything to test it? `opencv.__version__` yields 4.1.1.

**Comment by szk1509** [comment]
@hamddan4 well, i use (and recommend) you the official git repo... it includes also some tests and tutorials :)
https://docs.opencv.org/

**Pull request merged**